tests/test_query_model.py

```diff
@@ -132,6 +132,55 @@ def test_combining_non_patient_level_series_from_different_frames_throws_error()
         Function.EQ(event_date, vaccination_date)
 
 
+# You also can't combine results derived from the same source table if they've been
+# through divergent filter operations. Although such operations aren't *complete*
+# nonsense, there's never a good reason to do this and so instances of this are almost
+# certainly a result of user confusion or error and we want to reject them immediately.
+def test_combining_results_of_different_filter_operations_throws_an_error():
+    events = SelectTable("events", EVENTS_SCHEMA)
+    foo_events = Filter(events, Function.EQ(SelectColumn(events, "code"), Value("foo")))
+    bar_events = Filter(events, Function.EQ(SelectColumn(events, "code"), Value("bar")))
+    with pytest.raises(DomainMismatchError):
+        Function.EQ(SelectColumn(foo_events, "date"), SelectColumn(bar_events, "date"))
+
+
+# In particular you can't filter a frame using a predicate derived from a different
+# filter operation. Again, this isn't complete nonsense but it's not the best way to
+# express the query and is most probably the result of confusion.
+def test_filtering_frame_using_result_derived_from_another_filter_throws_error():
+    events = SelectTable("events", EVENTS_SCHEMA)
+    older_events = Filter(
+        events,
+        Function.LT(SelectColumn(events, "date"), Value(datetime.date(2022, 1, 1))),
+    )
+    recent_events = Filter(
+        events,
+        Function.GE(SelectColumn(events, "date"), Value(datetime.date(2022, 1, 1))),
+    )
+    is_recent_foo_event = Function.EQ(SelectColumn(recent_events, "code"), Value("foo"))
+    with pytest.raises(DomainMismatchError):
+        Filter(older_events, is_recent_foo_event)
+
+
+# However, it's fine to filter a Frame using a predicate derived from a direct ancenstor
+# of that Frame. This is important because it allows us to provide a "fluent" interface to
+# the Query Model. Without this, every time you wanted to filter a Frame you'd need a
+# reference to that Frame to use in the filter condition. And that would make it
+# impossible to use chained constructions like:
+#
+#     table.filter(...).filter(...).filter(...)
+#
+def test_filtering_frame_using_condition_derived_from_parent_frame_is_ok():
+    events = SelectTable("events", EVENTS_SCHEMA)
+    foo_events = Filter(events, Function.EQ(SelectColumn(events, "code"), Value("foo")))
+    recent_foo_events = Filter(
+        foo_events,
+        # Note that we're filtering `foo_events` using a condition derived from `events`
+        Function.GE(SelectColumn(events, "date"), Value(datetime.date(2022, 1, 1))),
+    )
+    assert recent_foo_events
+
+
```

Comment:
I think it would be helpful to have a test showing that combining non-patient level series with the same domain is okay.